modules/aws/test/src/smithy4s/aws/query/AwsQueryCodecSpec.scala

```diff
@@ -0,0 +1,354 @@
+      }
+    }
+
+    val xml = "foos.x.1=1&foos.x.2=2&foos.x.3=3"
+
+    checkContent(xml, Foo(List(1, 2, 3)))
+  }
+
+  test("list: flattened") {
+    case class Foo(foos: List[Int])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val foos = list(int)
+          .required[Foo]("foos", _.foos)
+          .addHints(XmlFlattened())
+        struct(foos)(Foo.apply).n
+      }
+    }
+    val xml = "foos.1=1&foos.2=2&foos.3=3"
+    checkContent(xml, Foo(List(1, 2, 3)))
+  }
+
+  test("list: flattened custom names") {
+    case class Foo(foos: List[Int])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val foos = list(int)
+          .required[Foo]("foos", _.foos)
+          .addHints(XmlFlattened(), XmlName("x"))
+        struct(foos)(Foo.apply).n
+      }
+    }
+    val xml = "x.1=1&x.2=2&x.3=3"
+
+    checkContent(xml, Foo(List(1, 2, 3)))
+  }
+
+  test("recursion") {
+    case class Foo(foo: Option[Foo])
+    object Foo {
+      implicit val schema: Schema[Foo] = recursive {
+        val foos = schema.optional[Foo]("foo", _.foo)
+        struct(foos)(Foo.apply).n
+      }
+    }
+
+    val xml = """|<Foo>
+                 |   <foo>
+                 |      <foo>
+                 |      </foo>
+                 |   </foo>
+                 |</Foo>
+                 |""".stripMargin
+    checkContent(xml, Foo(Some(Foo(Some(Foo(None))))))
+  }
+
+  test("union") {
+    type Foo = Either[Int, String]
+    implicit val schema: Schema[Foo] = {
+      val left = int.oneOf[Foo]("left", Left(_))
+      val right = string.oneOf[Foo]("right", Right(_))
+      union(left, right) {
+        case Left(int)     => left(int)
+        case Right(string) => right(string)
+      }
+    }
+    val xmlLeft = "left=1"
+    val xmlRight = "right=hello"
+    checkContent[Foo](xmlLeft, Left(1)) |+|
+      checkContent[Foo](xmlRight, Right("hello"))
+  }
+
+  test("union: custom names") {
+    type Foo = Either[Int, String]
+    implicit val schema: Schema[Foo] = {
+      val left = int.oneOf[Foo]("left", Left(_)).addHints(XmlName("foo"))
+      val right = string.oneOf[Foo]("right", Right(_)).addHints(XmlName("bar"))
+      union(left, right) {
+        case Left(int)     => left(int)
+        case Right(string) => right(string)
+      }
+    }
+    val xmlLeft = "foo=1"
+    val xmlRight = "bar=hello".stripMargin
+    checkContent[Foo](xmlLeft, Left(1)) |+|
+      checkContent[Foo](xmlRight, Right("hello"))
+  }
+
+  test("enumeration") {
+    sealed abstract class FooBar(val stringValue: String, val intValue: Int)
+        extends smithy4s.Enumeration.Value {
+      val name = stringValue
+      val value = stringValue
+      val hints = Hints.empty
+    }
+    object FooBar {
+      case object Foo extends FooBar("foo", 0)
+      case object Bar extends FooBar("bar", 1)
+      implicit val schema: Schema[FooBar] =
+        enumeration[FooBar](List(Foo, Bar)).x
+    }
+    val xmlFoo = "x=foo"
```

Comment:
this test is wrong : it should be `val xmlFoo = "foo"` (same for `xmlBar`)

Reply:
Done